File: apps/website/internal/services/token_purge.go
package services

import (
	"context"
	"log"
	"time"

	"gorm.io/gorm"

	"github.com/tbd-milsim/reforger-backend/internal/models"
)

// RefreshTokenRetention is the purge policy for refresh-token rows (T-130.1 F2B-09):
// a row is deleted once expires_at is more than this far in the past. Revoked rows
// must NOT be purged before expiry — they are the reuse-detection tripwire (a
// replayed spent token revokes the whole family, handlers Refresh, T-126 S2). Past
// expiry the token lookup rejects it regardless, so the extra week only buys clock
// skew tolerance and a forensic window.
const RefreshTokenRetention = 7 * 24 * time.Hour

// refreshPurgeInterval is how often StartRefreshTokenPurge re-sweeps after boot.
const refreshPurgeInterval = 6 * time.Hour

// PurgeExpiredRefreshTokens hard-deletes refresh-token rows that expired more than
// RefreshTokenRetention ago (RefreshToken has no soft-delete column) and returns
// the number of rows removed.
func PurgeExpiredRefreshTokens(db *gorm.DB) (int64, error) {
	res := db.Where("expires_at < ?", time.Now().Add(-RefreshTokenRetention)).
		Delete(&models.RefreshToken{})
	return res.RowsAffected, res.Error
}

// StartRefreshTokenPurge runs an immediate sweep and then re-sweeps every
// refreshPurgeInterval until ctx is cancelled. Failures are logged and retried on
// the next tick; the sweep never takes the API down.
func StartRefreshTokenPurge(ctx context.Context, db *gorm.DB) {
	go func() {
		sweep := func() {
			n, err := PurgeExpiredRefreshTokens(db)
			if err != nil {
				log.Printf("refresh token purge failed: %v", err)
				return
			}
			log.Printf("refresh token purge: %d rows", n)
		}
		sweep()
		ticker := time.NewTicker(refreshPurgeInterval)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				sweep()
			}
		}
	}()
}

File: apps/website/api_v2/src/services/token_purge.rs
//! Background refresh-token purge — Rust port of `services/token_purge.go`.
//!
//! Deletes refresh-token rows more than 7 days past expiry. Revoked-but-unexpired
//! rows are kept — they are the reuse-detection tripwire (see handlers::auth::refresh).

use chrono::{Duration, Utc};
use sqlx::PgPool;
use tokio::task::JoinHandle;

/// Retention past `expires_at` before a row is hard-deleted.
const RETENTION_DAYS: i64 = 7;
/// Re-sweep cadence after the immediate boot sweep.
const PURGE_INTERVAL: std::time::Duration = std::time::Duration::from_secs(6 * 60 * 60);

/// Handle to the background sweeper task (dropped/aborted on shutdown).
pub type PurgeHandle = JoinHandle<()>;

/// Hard-delete refresh-token rows that expired more than the retention window ago;
/// returns the number removed.
pub async fn purge_expired_refresh_tokens(pool: &PgPool) -> sqlx::Result<u64> {
    let cutoff = Utc::now() - Duration::days(RETENTION_DAYS);
    let res = sqlx::query("DELETE FROM refresh_tokens WHERE expires_at < $1")
        .bind(cutoff)
        .execute(pool)
        .await?;
    Ok(res.rows_affected())
}

/// Spawn the sweeper: an immediate sweep, then every 6h until the runtime stops.
pub fn start_refresh_token_purge(pool: PgPool) -> PurgeHandle {
    tokio::spawn(async move {
        sweep(&pool).await;
        let mut ticker = tokio::time::interval(PURGE_INTERVAL);
        ticker.tick().await; // consume the immediate first tick (already swept)
        loop {
            ticker.tick().await;
            sweep(&pool).await;
        }
    })
}

async fn sweep(pool: &PgPool) {
    match purge_expired_refresh_tokens(pool).await {
        Ok(n) => tracing::info!("refresh token purge: {n} rows"),
        Err(e) => tracing::error!(error = %e, "refresh token purge failed"),
    }
}
